{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioinformatics 2021: Week 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with BioPython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTCGATGCGATAGCCGTA\n",
      "||||| |||||.||||.||\n",
      "AGTCG-TGCGAAAGCCCTA\n",
      "  Score=28\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalms(\"AGTCGATGCGATAGCCGTA\", \"AGTCGTGCGAAAGCCCTA\", 2, -1.5, -1, -0.5)\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('AGTCGATGCGATAGCCGTA', 'AGTCG-TGCGAAAGCCCTA', 28.0)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments[0].seqA, alignments[0].seqB, alignments[0].score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install dendropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dendropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://dendropy.org/primer/trees.html#building-a-tree-programmatically\n",
    "\n",
    "taxon_namespace = dendropy.TaxonNamespace([\"A\", \"B\", \"C\", \"D\"])\n",
    "tree = dendropy.Tree(taxon_namespace=taxon_namespace)\n",
    "\n",
    "ch1 = tree.seed_node.new_child()\n",
    "ch1.edge.length = 1\n",
    "ch2 = tree.seed_node.new_child(edge_length=3)\n",
    "\n",
    "ch3 = dendropy.Node(edge_length=1)\n",
    "ch4 = dendropy.Node(edge_length=2)\n",
    "ch1.add_child(ch3)\n",
    "ch1.add_child(ch4)\n",
    "\n",
    "ch5 = dendropy.Node(edge_length=4)\n",
    "ch6 = dendropy.Node(edge_length=5)\n",
    "\n",
    "ch3.set_child_nodes([ch5, ch6])\n",
    "\n",
    "# Assign taxa\n",
    "ch2.taxon = taxon_namespace.get_taxon(\"A\")\n",
    "ch4.taxon = taxon_namespace.get_taxon(\"B\")\n",
    "ch5.taxon = taxon_namespace.get_taxon(\"C\")\n",
    "ch6.taxon = taxon_namespace.get_taxon(\"D\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                             /-------------------------------------- C\n",
      "                                      /--------------------------------------+                                        \n",
      "/-------------------------------------+                                      \\-------------------------------------- D\n",
      "|                                     |                                                                               \n",
      "+                                     \\----------------------------------------------------------------------------- B\n",
      "|                                                                                                                     \n",
      "\\------------------------------------------------------------------------------------------------------------------- A\n",
      "                                                                                                                      \n",
      "                                                                                                                      \n"
     ]
    }
   ],
   "source": [
    "# visual representation\n",
    "print(tree.as_ascii_plot())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(((C:4,D:5):1,B:2):1,A:3);\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Newick format \n",
    "print(tree.as_string(\"newick\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with numpy "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "arr = np.ndarray((3,3))\n",
    "for i in range(3):\n",
    "    for j in range(3):\n",
    "        arr[i,j] = (i+1)+(j+1)*10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[11., 21., 31.],\n",
       "       [12., 22., 32.],\n",
       "       [13., 23., 33.]])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tasks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Task 1: construct pairwise distance matrix for a list of sequences:\n",
    "# - align two sequences\n",
    "# - use hamming_distance\n",
    "\n",
    "def distance_matrix_pairwise(seqs):\n",
    "    dm = np.ndarray((len(seqs),len(seqs)))\n",
    "    # your code\n",
    "    \n",
    "    return dm\n",
    "\n",
    "# Example 1: pairwise_distance([\"AA\", \"AC\", \"GC\"]) -> \n",
    "#                                                    [[0,1,2],\n",
    "#                                                     [1,0,1],\n",
    "#                                                     [2,1,0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Task 2: perform hierarchical clustering\n",
    "\n",
    "def hierarchical_clustering(dm, taxa):\n",
    "    taxon_namespace = dendropy.TaxonNamespace(taxa)\n",
    "    tree = dendropy.Tree(taxon_namespace=taxon_namespace)\n",
    "    # your code\n",
    "    \n",
    "    return tree\n",
    "\n",
    "# Inputs:\n",
    "# - NxN distance matrix for N objects\n",
    "# - list of names for N objects\n",
    "\n",
    "# Brief notes:\n",
    "# - two nearest objects are joined into a cluster\n",
    "# - the corresponding edge_length should be taken as their respective distance / 2\n",
    "# - algorithm iterates until a signle root cluster is formed\n",
    "# - distance between clusters should be calculated as mean pairwise distance \n",
    "#   between all objects in the 1st cluster and all objects in the 2nd cluster (or suggest your idea?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Task X: compare with scipy \n",
    "# https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html\n",
    "# consider different methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from scipy.spatial.distance import pdist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'icoord': [[25.0, 25.0, 35.0, 35.0],\n",
       "  [15.0, 15.0, 30.0, 30.0],\n",
       "  [5.0, 5.0, 22.5, 22.5]],\n",
       " 'dcoord': [[0.0, 3.1622776601683795, 3.1622776601683795, 0.0],\n",
       "  [0.0, 3.872983346207417, 3.872983346207417, 3.1622776601683795],\n",
       "  [0.0, 5.744562646538029, 5.744562646538029, 3.872983346207417]],\n",
       " 'ivl': ['C', 'D', 'A', 'B'],\n",
       " 'leaves': [2, 3, 0, 1],\n",
       " 'color_list': ['C1', 'C1', 'C0'],\n",
       " 'leaves_color_list': ['C0', 'C1', 'C1', 'C1']}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAD+CAYAAAATWE8CAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAALnklEQVR4nO3da4xcBR2G8fe1xQuKojJSQqmriTESK2BWjFFjxKgoKHwgXvBCqslqogYv8fbFROMtfvAab41aTQCVVBCViJIoKhHQrVZXWkyUoFJZmXqDEm+trx9mCm1dmNNlzp7/7Dy/ZLI7u9PlzWZ59vTsma2TCABQ1326HgAAuGeEGgCKI9QAUByhBoDiCDUAFEeoAaC4RqG2fbTtrbZvsL3T9lPaHgYAGFjb8HEfk3RFknNs31fSkS1uAgAcwKOe8GL7IZK2S3p0Gj475phjjsnMzMy9HgcA02Lbtm27k/SWel+TI+pHSepL2mL7JEnbJJ2f5I67+wMzMzOan59f1lgAmEa2f3d372tyjnqtpCdK+nSSUyTdIekdS/xH5mzP257v9/vLHgsAOFiTUN8s6eYk1w3vb9Ug3AdJsjnJbJLZXm/Jo3cAwDKMDHWSRUl/sP3Y4ZueJWlHq6sAAHdqetXHGyRdOLzi40ZJm9qbBAA4UKNQJ9kuabbdKQCApfDMRAAojlADQHGEGgCKa/rDxFXlout+r8u27+p6BlaJs04+Xuc+eUPXM7CKTeUR9WXbd2nHLbd1PQOrwI5bbuObPlo3lUfUknTicQ/WV1/DLwHEvfPiz17T9QRMgak8ogaASUKoAaA4Qg0AxRFqACiOUANAcYQaAIoj1ABQHKEGgOIINQAUR6gBoDhCDQDFEWoAKI5QA0BxhBoAiiPUAFAcoQaA4gg1ABRHqAGgOEINAMURagAojlADQHGN/hVy2zdJul3SPkl7k8y2OQoAcJdGoR56ZpLdrS0BACyJUx8AUFzTUEfSd21vsz3X5iAAwMGanvp4WpJdth8h6UrbNyT54YEPGAZ8TpI2bNgw5pkAML0aHVEn2TV8eaukSyWdusRjNieZTTLb6/XGuxIAptjIUNt+oO2j9r8u6TmSftX2MADAQJNTH8dKutT2/sdflOSKVlcBAO40MtRJbpR00gpsAQAsgcvzAKA4Qg0AxRFqACiOUANAcYQaAIoj1ABQHKEGgOIINQAUR6gBoDhCDQDFEWoAKI5QA0BxhBoAiiPUAFAcoQaA4gg1ABRHqAGgOEINAMURagAojlADQHGEGgCKI9QAUByhBoDiCDUAFEeoAaA4Qg0AxRFqACiucahtr7H9c9vfanMQAOBgh3NEfb6knW0NAQAsrVGoba+XdIakz7U7BwBwqKZH1B+V9DZJ/21vCgBgKSNDbftMSbcm2TbicXO2523P9/v9sQ0EgGnX5Ij6qZJeaPsmSV+RdJrtCw59UJLNSWaTzPZ6vTHPBIDpNTLUSd6ZZH2SGUkvkfS9JC9vfRkAQJK0tusBWAXmt0gLW7te0Y3FswYvt7y32x1d2XiONLup6xWr3mGFOslVkq5qZQkm18JWaXFBWrex6yUr7qsbLut6QncWFwYvCXXrOKLGeKzbKG26vOsVWElbzuh6wdTgKeQAUByhBoDiCDUAFEeoAaA4Qg0AxRFqACiOUANAcYQaAIoj1ABQHKEGgOIINQAUR6gBoDhCDQDFEWoAKI5QA0BxhBoAiiPUAFAcoQaA4gg1ABRHqAGgOEINAMURagAojlADQHFrux4A4BDzW6SFrV2vGG3xl4OXW87odkcTG8+RZjd1vWLZOKIGqlnYKi0udL1itHVPGNyqW1yYjG9894AjaqCidRulTZd3vWJ1mIQj/hFGHlHbvr/tn9j+he3rbb97JYYBAAaaHFH/S9JpSfbYPkLS1ba/neTalrcBANQg1Ekiac/w7hHDW9ocBQC4S6MfJtpeY3u7pFslXZnkulZXAQDu1CjUSfYlOVnSekmn2n78oY+xPWd73vZ8v98f80wAmF6HdXlekr9J+r6k05d43+Yks0lme73emOYBAJpc9dGzffTw9QdIerakG1reBQAYanLVx3GSvmR7jQZhvzjJt9qdBQDYr8lVH7+UdMoKbAEALIGnkANAcYQaAIoj1ABQHKEGgOIINQAUR6gBoDhCDQDFEWoAKI5QA0BxhBoAiiPUAFAcoQaA4gg1ABRHqAGgOEINAMURagAojlADQHGEGgCKI9QAUByhBoDiCDUAFEeoAaA4Qg0AxRFqACiOUANAcYQaAIoj1ABQ3MhQ2z7B9vdt77B9ve3zV2IYAGBgbYPH7JX0liQ/s32UpG22r0yyo+VtAAA1OKJOckuSnw1fv13STknHtz0MADBwWOeobc9IOkXSda2sAQD8n8ahtv0gSV+T9MYkty3x/jnb87bn+/3+ODcCwFRrFGrbR2gQ6QuTXLLUY5JsTjKbZLbX641zIwBMtSZXfVjS5yXtTPLh9icBAA7U5Ij6qZJeIek029uHt+e3vAsAMDTy8rwkV0vyCmwBACyBZyYCQHGEGgCKI9QAUByhBoDiCDUAFEeoAaA4Qg0AxRFqACiOUANAcYQaAIoj1ABQHKEGgOIINQAUR6gBoDhCDQDFEWoAKI5QA0BxhBoAiiPUAFAcoQaA4gg1ABRHqAGgOEINAMURagAojlADQHGEGgCKI9QAUNzIUNv+gu1bbf9qJQYBAA7W5Ij6i5JOb3kHAOBujAx1kh9K+ssKbAEALIFz1ABQ3NhCbXvO9rzt+X6/P64PCwBTb2yhTrI5yWyS2V6vN64PCwBTj1MfAFBck8vzvizpGkmPtX2z7Ve3PwsAsN/aUQ9I8tKVGAIAWBqnPgCgOEINAMURagAojlADQHGEGgCKI9QAUByhBoDiCDUAFEeoAaA4Qg0AxRFqACiOUANAcYQaAIoj1ABQHKEGgOIINQAUR6gBoDhCDQDFEWoAKI5QA0BxhBoAiiPUAFAcoQaA4gg1ABRHqAGgOEINAMURagAorlGobZ9u+9e2f2P7HW2PAgDcZWSoba+R9ElJz5N0oqSX2j6x7WEAgIEmR9SnSvpNkhuT/FvSVySd1e4sAMB+TUJ9vKQ/HHD/5uHbAAArYO24PpDtOUlzw7t7bP96XB+7LRe/tusFq8yr3PWC1YXP53jV/3w+8u7e0STUuySdcMD99cO3HSTJZkmbD3saAOAeNTn18VNJj7H9KNv3lfQSSd9odxYAYL+RR9RJ9tp+vaTvSFoj6QtJrm99GQBAkuQkXW8AANwDnpkIAMURagAojlADQHFTGWrb59qet73H9i22v237aV3vmjS2b7L9D9u32/6b7R/bfq3tqfy6GhfbV9n+q+37db1lkh3w9bln+Pm83PYJo/9kPVP3P5TtN0v6qKT3SzpW0gZJnxJPi1+uFyQ5SoOL9T8o6e2SPt/tpMlle0bS0yVF0gu7XbMqvCDJgyQdJ+lPkj7R8Z5lmapQ236IpPdIel2SS5LckeQ/Sb6Z5K1d75tkSf6e5BuSXizpPNuP73rThHqlpGslfVHSed1OWT2S/FPSVg1+sdzEmapQS3qKpPtLurTrIatVkp9o8Ptgnt71lgn1SkkXDm/PtX1sx3tWBdtHanAQcW3XW5ZjbL/rY0I8XNLuJHu7HrLK/VHSw7oeMWmGPyd5pKSLk+y2/VtJ50r6SLfLJtrXbe+V9EBJfUnP7XjPskzbEfWfJR1je9q+Qa204yX9pesRE+g8Sd9Nsnt4/yJx+uPeOjvJ0Rr8Tfr1kn5ge123kw7ftIX6Gkn/knR2xztWLdtP0iDUV3e9ZZLYfoCkF0l6hu1F24uS3iTpJNsndbtu8iXZl+QSSfskTdwVXlMV6iR/l/QuSZ+0fbbtI20fYft5tj/U9b5JZvvBts/U4B+WuCDJQtebJszZGkTkREknD2+Pk/QjDc5b417wwFmSHippZ9d7DtdU/q4P2y/T4GjlcZJul7RN0vuS/LjTYRPG9k0aXOK4V9J/Je2QdIGkzyTZ1+G0iWP7CknXJ3nLIW9/kaSPS1rPz1YOzwFfn/s0uNzxd5I+kOTCLnctx1SGGgAmyVSd+gCASUSoAaA4Qg0AxRFqACiOUANAcYQaAIoj1ABQHKEGgOIINQAU9z8hhlwG1ggBCQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "taxa = [\"A\", \"B\", \"C\", \"D\"]\n",
    "dm = np.array([[0,1,4,3],\n",
    "               [1,0,2,1],\n",
    "               [4,2,0,5],\n",
    "               [3,1,5,0]])\n",
    "tr = linkage(pdist(dm))\n",
    "dendrogram(tr, labels=taxa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
